packages/shared/src/hooks/useDefaultFeed.ts

```diff
@@ -1,14 +1,51 @@
+import { useEffect } from 'react';
+import { MainFeedPage } from '../components/utilities';
 import usePersistentContext from './usePersistentContext';
 
-export default function useDefaultFeed(): [
-  string,
-  (value: string) => Promise<void>,
-] {
+interface UseDefaultFeedProps {
+  hasFiltered?: boolean;
+  hasUser?: boolean;
+  feed?: string;
+}
+
+export const getShouldRedirect = (
+  isOnMyFeed: boolean,
+  isLoggedIn: boolean,
+): boolean => {
+  if (!isOnMyFeed) {
+    return false;
+  }
+
+  if (!isLoggedIn) {
+    return true;
+  }
+
+  return false;
+};
+
+export default function useDefaultFeed({
+  hasFiltered,
+  hasUser,
+  feed,
+}: UseDefaultFeedProps = {}): string {
+  const isMyFeed = feed === MainFeedPage.MyFeed;
   const [defaultFeed, updateDefaultFeed] = usePersistentContext(
     'defaultFeed',
-    'my-feed',
+    !hasUser || !hasFiltered ? 'popular' : 'my-feed',
     ['my-feed', 'popular', 'upvoted', 'discussed'],
+    !hasUser || !hasFiltered ? 'popular' : 'my-feed',
   );
 
-  return [defaultFeed, updateDefaultFeed];
+  useEffect(() => {
+    const feedConditions = [null, defaultFeed, 'default', '/'];
+    if (
+      defaultFeed !== null &&
+      feedConditions.every((condition) => condition !== feed) &&
+      !getShouldRedirect(isMyFeed, !!hasUser)
+    ) {
+      updateDefaultFeed(feed);
+    }
+  }, [defaultFeed, feed]);
+
+  return defaultFeed;
```

Comment:
Perhaps we should memo this?

Reply:
I thought the `usePersistentContext` was memoizing the value already, but after checking, it seems not. Let me explicitly memoize this quickly.